# Contexto de los datos



Tenemos los datos de los usuarios en la plataforma YELP, y las diferentes columnas representan:

1. __user_id__: identificador unico del usuario.

2. __name__ : nombre.

3. __review_count__ : el numero de reseñas que ha realizados.

4. __yelping_since__ : fecha en que se unio el usuario.

De las reseñas realizadas por el usuario cuantas fueron marcadas:

5. __useful__ :  útiles.

6. __funny__ : divertidas.

7. __cool__ :  geniales.

Procedemos con las demas columnas:

8. __elite__ : información si el usuario alcanzo el estado elite.

9. __friends__ : lista de amigos del usuario.

10. __fans__ : número de seguidores del usuario.

11. __average_stars__ : el promedio de calificaciones de estrelas que el usuario ha dado en sus reseñas.

Las columnas que comienzan compliment: reflejan a las valoraciones que recibio el usuario sobre sus contribuciones a la plataforma, que pueden incluir reseñas, fotos comentarios u otros.

12. __compliment__:

# Importando Librerias

In [ ]:
import pandas as pd
from google.colab import drive




# Montar Google Drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import matplotlib.pyplot as plt
import numpy as np


# Cargando datos

Obtenemos el dataframe: registros_unicos

### Panomara General del dataframe

In [ ]:
data = pd.read_parquet("/content/drive/MyDrive/Proyecto Grupal/user.parquet")

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2105597 entries, 0 to 2105596
Data columns (total 22 columns):
 #   Column              Dtype  
---  ------              -----  
 0   user_id             object 
 1   name                object 
 2   review_count        int64  
 3   yelping_since       object 
 4   useful              int64  
 5   funny               int64  
 6   cool                int64  
 7   elite               object 
 8   friends             object 
 9   fans                int64  
 10  average_stars       float64
 11  compliment_hot      int64  
 12  compliment_more     int64  
 13  compliment_profile  int64  
 14  compliment_cute     int64  
 15  compliment_list     int64  
 16  compliment_note     int64  
 17  compliment_plain    int64  
 18  compliment_cool     int64  
 19  compliment_funny    int64  
 20  compliment_writer   int64  
 21  compliment_photos   int64  
dtypes: float64(1), int64(16), object(5)
memory usage: 353.4+ MB


In [ ]:
data.head()

user_id    name  review_count        yelping_since  useful  \
0  qVc8ODYU5SZjKXVBgXdI7w  Walker           585  2007-01-25 16:47:26    7217   
1  j14WgRoU_-2ZE1aw1dXrJg  Daniel          4333  2009-01-25 04:35:42   43091   
2  2WnXYQFK0hXEoTxPtV2zvg   Steph           665  2008-07-25 10:41:00    2086   
3  SZDeASXq7o05mMNLshsdIA    Gwen           224  2005-11-29 04:38:33     512   
4  hA5lMy-EnncsH4JoR-hFGQ   Karen            79  2007-01-05 19:40:59      29   

   funny   cool                                              elite  \
0   1259   5994                                               2007   
1  13066  27281  2009,2010,2011,2012,2013,2014,2015,2016,2017,2...   
2   1010   1003                           2009,2010,2011,2012,2013   
3    330    299                                     2009,2010,2011   
4     15      7                                                      

                                             friends  fans  ...  \
0  NSCy54eWehBJyZdG2iE84w, pe42u7DcCH2QmI81NX-8qA...   267  ...   
1  ueRPE0CX75ePGMqOFVj6IQ, 52oH4DrRvzzl8wh5UXyU0A...  3138  ...   
2  LuO3Bn4f3rlhyHIaNfTlnA, j9B4XdHUhDfTKVecyWQgyA...    52  ...   
3  enx1vVPnfdNUdPho6PH_wg, 4wOcvMLtU6a9Lslggq74Vg...    28  ...   
4  PBK4q9KEEBHhFvSXCUirIw, 3FWPpM7KU1gXeOM_ZbYMbA...     1  ...   

   compliment_more  compliment_profile  compliment_cute  compliment_list  \
0               65                  55               56               18   
1              264                 184              157              251   
2               13                  10               17                3   
3                4                   1                6                2   
4                1                   0                0                0   

   compliment_note  compliment_plain  compliment_cool  compliment_funny  \
0              232               844              467               467   
1             1847              7054             3131              3131   
2               66                96              119               119   
3               12                16               26                26   
4                1                 1                0                 0   

   compliment_writer  compliment_photos  
0                239                180  
1               1521               1946  
2                 35                 18  
3                 10                  9  
4                  0                  0  

[5 rows x 22 columns]

### Valores Nulos

In [ ]:
valores_nulos = data.isnull().sum()

In [ ]:
valores_nulos

user_id               0
name                  0
review_count          0
yelping_since         0
useful                0
funny                 0
cool                  0
elite                 0
friends               0
fans                  0
average_stars         0
compliment_hot        0
compliment_more       0
compliment_profile    0
compliment_cute       0
compliment_list       0
compliment_note       0
compliment_plain      0
compliment_cool       0
compliment_funny      0
compliment_writer     0
compliment_photos     0
dtype: int64

### Valores unicos de cada columna

In [ ]:
# Encuentra los valores únicos en la columna

columna = 'user_id'

cantidad_valores_unicos = data[columna].nunique()

print(cantidad_valores_unicos)

1987897


### Eliminando los duplicados de la columna user_id

Esto lo realizamos debido a que tenemos:

valores unicos = 1987897; registros totales = 2105597

Realizar la eliminacion de los registros duplicados, representaria el 5%, lo cual podemos concluir que no representa una cantidad considerable del data set original.

In [ ]:
# Usar el método 'duplicated' para identificar registros duplicados en 'id'
duplicados = data[columna].duplicated(keep=False)

# Filtrar el DataFrame para mantener solo los registros no duplicados
registros_unicos = data[~duplicados]


In [ ]:
registros_unicos = registros_unicos.reset_index(drop=True)


In [ ]:
registros_unicos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1944597 entries, 0 to 1944596
Data columns (total 22 columns):
 #   Column              Dtype  
---  ------              -----  
 0   user_id             object 
 1   name                object 
 2   review_count        int64  
 3   yelping_since       object 
 4   useful              int64  
 5   funny               int64  
 6   cool                int64  
 7   elite               object 
 8   friends             object 
 9   fans                int64  
 10  average_stars       float64
 11  compliment_hot      int64  
 12  compliment_more     int64  
 13  compliment_profile  int64  
 14  compliment_cute     int64  
 15  compliment_list     int64  
 16  compliment_note     int64  
 17  compliment_plain    int64  
 18  compliment_cool     int64  
 19  compliment_funny    int64  
 20  compliment_writer   int64  
 21  compliment_photos   int64  
dtypes: float64(1), int64(16), object(5)
memory usage: 326.4+ MB


# Análisis de las variables

In [ ]:
registros_unicos.describe()

review_count        useful         funny          cool          fans  \
count  1.944544e+06  1.944544e+06  1.944544e+06  1.944544e+06  1.944544e+06   
mean   2.123474e+01  3.641283e+01  1.423948e+01  2.015345e+01  1.258229e+00   
std    6.976534e+01  5.828789e+02  3.744473e+02  5.219346e+02  1.368852e+01   
min    1.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    2.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    5.000000e+00  3.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    1.600000e+01  1.200000e+01  2.000000e+00  3.000000e+00  0.000000e+00   
max    1.747300e+04  2.062960e+05  1.858230e+05  1.998780e+05  3.642000e+03   

       average_stars  compliment_hot  compliment_more  compliment_profile  \
count   1.944544e+06    1.944544e+06     1.944544e+06        1.944544e+06   
mean    3.625963e+00    1.484755e+00     2.425792e-01        1.421773e-01   
std     1.192286e+00    7.001382e+01     1.207230e+01        1.357798e+01   
min     1.000000e+00    0.000000e+00     0.000000e+00        0.000000e+00   
25%     3.000000e+00    0.000000e+00     0.000000e+00        0.000000e+00   
50%     3.880000e+00    0.000000e+00     0.000000e+00        0.000000e+00   
75%     4.570000e+00    0.000000e+00     0.000000e+00        0.000000e+00   
max     5.000000e+00    2.578400e+04     1.350100e+04        1.418000e+04   

       compliment_cute  compliment_list  compliment_note  compliment_plain  \
count     1.944544e+06     1.944544e+06     1.944544e+06      1.944544e+06   
mean      1.061853e-01     4.663510e-02     1.213911e+00      2.471001e+00   
std       1.112663e+01     9.749519e+00     5.914592e+01      1.103870e+02   
min       0.000000e+00     0.000000e+00     0.000000e+00      0.000000e+00   
25%       0.000000e+00     0.000000e+00     0.000000e+00      0.000000e+00   
50%       0.000000e+00     0.000000e+00     0.000000e+00      0.000000e+00   
75%       0.000000e+00     0.000000e+00     0.000000e+00      0.000000e+00   
max       1.365400e+04     1.266900e+04     5.903100e+04      1.010970e+05   

       compliment_cool  compliment_funny  compliment_writer  \
count     1.944544e+06      1.944544e+06       1.944544e+06   
mean      2.349343e+00      2.349343e+00       8.622191e-01   
std       9.145991e+01      9.145991e+01       2.943743e+01   
min       0.000000e+00      0.000000e+00       0.000000e+00   
25%       0.000000e+00      0.000000e+00       0.000000e+00   
50%       0.000000e+00      0.000000e+00       0.000000e+00   
75%       0.000000e+00      0.000000e+00       0.000000e+00   
max       4.996700e+04      4.996700e+04       1.593400e+04   

       compliment_photos  review_count_normalized  
count       1.944544e+06             1.944544e+06  
mean        1.093474e+00             1.158124e-03  
std         9.439275e+01             3.992980e-03  
min         0.000000e+00             0.000000e+00  
25%         0.000000e+00             5.723443e-05  
50%         0.000000e+00             2.289377e-04  
75%         0.000000e+00             8.585165e-04  
max         8.263000e+04             1.000000e+00

In [ ]:
registros_unicos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1944597 entries, 0 to 1944596
Data columns (total 22 columns):
 #   Column              Dtype  
---  ------              -----  
 0   user_id             object 
 1   name                object 
 2   review_count        int64  
 3   yelping_since       object 
 4   useful              int64  
 5   funny               int64  
 6   cool                int64  
 7   elite               object 
 8   friends             object 
 9   fans                int64  
 10  average_stars       float64
 11  compliment_hot      int64  
 12  compliment_more     int64  
 13  compliment_profile  int64  
 14  compliment_cute     int64  
 15  compliment_list     int64  
 16  compliment_note     int64  
 17  compliment_plain    int64  
 18  compliment_cool     int64  
 19  compliment_funny    int64  
 20  compliment_writer   int64  
 21  compliment_photos   int64  
dtypes: float64(1), int64(16), object(5)
memory usage: 326.4+ MB


In [ ]:
registros_unicos.head()

user_id     name  review_count        yelping_since  useful  \
0  LhfpZHG5fcsMtKzkR7cygQ   Angela            32  2008-12-08 14:46:35      85   
1  7jRqME7-krCDvdkR-dadgA  Michael           218  2008-08-20 02:49:12     594   
2  h6pfiioTL2ypQt7CMZEgaw   Andrew           196  2008-03-15 01:36:14     478   
3  JLWan2EBO9YQi8nf1PEA5A   Martin            10  2009-01-19 16:37:07      13   
4  85uM-cor3VXCUyVFEe0YNg     Matt           573  2009-09-14 13:29:47    1096   

   funny  cool                                    elite  \
0     23    46                                            
1    383   189  2009,2010,2011,2012,2013,2014,2015,2016   
2    231   263                                2009,2010   
3      2     7                                            
4    195   347  2010,2011,2012,2013,2014,2015,2016,2017   

                                             friends  fans  ...  \
0  -4kMeC1vaCIT-PWXUA4PrA, Xurdy-KOvYv4lTybXXHxTQ...     7  ...   
1  JIViJRj6uhVqB7wo5q_udg, hBicl52tMVsTecdZSVL-lQ...    12  ...   
2  vASaF47nWvCHEEDA8nfCVw, 4i4lyXBigT2HShIjw7TbDw...     7  ...   
3  dIjX8WVl9UMawuXJzi_gtQ, ya5ALbF0OZEAiVi7wf3KoQ...     0  ...   
4  4Y3Xy1nG2QzwgDLNcIeAIw, FrOWaaFLUhmD4ysysNDzaA...    22  ...   

   compliment_more  compliment_profile  compliment_cute  compliment_list  \
0                1                   0                0                0   
1                5                   2                0                1   
2                4                   3               10                0   
3                0                   0                0                0   
4                6                   2                2                1   

   compliment_note  compliment_plain  compliment_cool  compliment_funny  \
0                2                 3                1                 1   
1               18                 5               14                14   
2                8                36               29                29   
3                0                 0                1                 1   
4               16                37               45                45   

   compliment_writer  compliment_photos  
0                  2                  0  
1                  2                  2  
2                 10                  3  
3                  0                  0  
4                 26                  8  

[5 rows x 22 columns]


### Analizando la columna review_count

In [ ]:
# Encontrar el valor máximo en la columna 'review_count'
max_review_count = registros_unicos['review_count'].max()

print("El valor máximo en la columna 'review_count' es:", max_review_count)


El valor máximo en la columna 'review_count' es: 17473


In [ ]:
# Buscar el registro con 'review_count'
registro_deseado = registros_unicos.loc[registros_unicos['review_count'] == 17473]

# Imprimir el registro encontrado
registro_deseado

user_id name  review_count        yelping_since  useful  \
6069  Hi10sGSZNxQH3NLyWSZ1oA  Fox         17473  2009-05-26 11:33:58  206296   

       funny    cool                                     elite  \
6069  185823  195814  2014,2015,2016,2017,2018,2019,20,20,2021   

                                                friends  fans  ...  \
6069  uZcWE40gWJHjLT_VyNBp1Q, N52nbyDKZte4PUfJVNpkGQ...  3493  ...   

      compliment_more  compliment_profile  compliment_cute  compliment_list  \
6069              431                 402              130               45   

      compliment_note  compliment_plain  compliment_cool  compliment_funny  \
6069             2408              6565             7655              7655   

      compliment_writer  compliment_photos  
6069               1644               4053  

[1 rows x 22 columns]

In [ ]:
# Contar los registros que tienen valores iguales a cero en 'review_count'
count_zeros = (registros_unicos['review_count'] == 0).sum()

print(f"La cantidad de registros con valores iguales a cero en 'review_count' es: {count_zeros}")


La cantidad de registros con valores iguales a cero en 'review_count' es: 0


#### Realizamos la eliminación de los registros con review_count =0. Esto se realiza debido a que no tiene interaccion con la pltaforma.

In [ ]:
# Eliminar los registros con valor igual a cero en 'review_count'
registros_unicos = registros_unicos[registros_unicos['review_count'] != 0]


#### Normalizamos la columna review_count en otra columna de nombre review_count_normalized para posteiormente visualizar

In [ ]:
# Calcular el mínimo y el máximo de la columna 'review_count'
min_review_count = registros_unicos['review_count'].min()
max_review_count = registros_unicos['review_count'].max()

# Aplicar la normalización Min-Max a la columna 'review_count'
registros_unicos['review_count_normalized'] = (registros_unicos['review_count'] - min_review_count) / (max_review_count - min_review_count)

In [ ]:
registros_unicos.head()

user_id     name  review_count        yelping_since  useful  \
0  LhfpZHG5fcsMtKzkR7cygQ   Angela            32  2008-12-08 14:46:35      85   
1  7jRqME7-krCDvdkR-dadgA  Michael           218  2008-08-20 02:49:12     594   
2  h6pfiioTL2ypQt7CMZEgaw   Andrew           196  2008-03-15 01:36:14     478   
3  JLWan2EBO9YQi8nf1PEA5A   Martin            10  2009-01-19 16:37:07      13   
4  85uM-cor3VXCUyVFEe0YNg     Matt           573  2009-09-14 13:29:47    1096   

   funny  cool                                    elite  \
0     23    46                                            
1    383   189  2009,2010,2011,2012,2013,2014,2015,2016   
2    231   263                                2009,2010   
3      2     7                                            
4    195   347  2010,2011,2012,2013,2014,2015,2016,2017   

                                             friends  fans  ...  \
0  -4kMeC1vaCIT-PWXUA4PrA, Xurdy-KOvYv4lTybXXHxTQ...     7  ...   
1  JIViJRj6uhVqB7wo5q_udg, hBicl52tMVsTecdZSVL-lQ...    12  ...   
2  vASaF47nWvCHEEDA8nfCVw, 4i4lyXBigT2HShIjw7TbDw...     7  ...   
3  dIjX8WVl9UMawuXJzi_gtQ, ya5ALbF0OZEAiVi7wf3KoQ...     0  ...   
4  4Y3Xy1nG2QzwgDLNcIeAIw, FrOWaaFLUhmD4ysysNDzaA...    22  ...   

   compliment_profile  compliment_cute  compliment_list  compliment_note  \
0                   0                0                0                2   
1                   2                0                1               18   
2                   3               10                0                8   
3                   0                0                0                0   
4                   2                2                1               16   

   compliment_plain  compliment_cool  compliment_funny  compliment_writer  \
0                 3                1                 1                  2   
1                 5               14                14                  2   
2                36               29                29                 10   
3                 0                1                 1                  0   
4                37               45                45                 26   

   compliment_photos  review_count_normalized  
0                  0                 0.001774  
1                  2                 0.012420  
2                  3                 0.011161  
3                  0                 0.000515  
4                  8                 0.032738  

[5 rows x 23 columns]

### Analizando la columna yelping

In [ ]:
# Puedes verificar los valores nulos en esa columna de la siguiente manera:
null_values = registros_unicos['yelping_since'].isnull()

# Ahora, puedes imprimir los registros que tienen valores nulos en la columna 'yelping_since'
print(registros_unicos[null_values])

Empty DataFrame
Columns: [user_id, name, review_count, yelping_since, useful, funny, cool, elite, friends, fans, average_stars, compliment_hot, compliment_more, compliment_profile, compliment_cute, compliment_list, compliment_note, compliment_plain, compliment_cool, compliment_funny, compliment_writer, compliment_photos, review_count_normalized]
Index: []

[0 rows x 23 columns]


#### Se realiza el cambio de tipo de dato

In [ ]:
# Suponiendo que tienes un DataFrame llamado 'df' y quieres convertir la columna 'yelping_since' a datetime64
registros_unicos['yelping_since'] = pd.to_datetime(registros_unicos['yelping_since'])